### ISSUE TITLE
Addition of Rotational Motion State Derivatives

### ISSUE THREAD

**Issue by mfacchinelli** [Tudat/tudat#385]
Hi Dominic, 

I realized that I had forgotten to add the rotational state derivatives for MRP and exponential map. I have started doing it, but it would be nice if you could take a look at what I have changed, and tell me if I have missed some things. I have pushed the latest changes to:

https://github.com/mfacchinelli/tudat/tree/rotationalMotion

In particular, I am not sure if what I have done so far is enough to 'connect' the new propagators to the main simulation body. Also, I don't really know what should be done to the state inside the propagator-specific constructors, such as `RotationalMotionQuaternionsStateDerivative`.  Right now it's empty, since the previous lines of code are in the more general `RotationalMotionStateDerivative` class.

Right now the code compiles, and when the unit test on rotational motion I get only a mere 34503 failures 😅 So I would guess something goes wrong. I will take a look at why that happens tomorrow. 

Thank you!

Have a nice weekend, 

Michele

**Comment by DominicDirkx** [comment]
Hi Michele,

This is (yet another :) ) great initiative. Making the rotational dynamics propagation more flexible is a great idea, and what you have done seems to mirror the approach of the NBodyStateDerivative stuff, which is the right way to go, I think.

I assume that the RotationalMotionQuaternionsStateDerivative should have the functionality that the 'old' setup had. In principle, anything that is specific to this state derivative model should be moved to the RotationalMotionQuaternionsStateDerivative class. Only those aspects that are common to all rotational state derivative models should be placed in the updated RotationalMotionStateDerivative class. 

Let me know if you get stuck debugging anything. 

Have a nice weekend,

Dominic

**Comment by mfacchinelli** [comment]
Hi Dominic, 

Right now I am struggling with the `convertToOutputSolution` function. In the description it says that the converted state is 'returned by reference', but when I use the function `getEquationsOfMotionNumericalSolution`, it returns the 'local' solution (e.g., for exponential map it outputs the exponential map state, instead of quaternions). How and where does it save the converted result? 

Another issue, is that when I use quaternions or MRPs, the output of the *translational* state from `getEquationsOfMotionNumericalSolution`, becomes very strange:
```
 3.17537e+06  1.45172e+06 -1.35912e+06      1188.48       335.29      3134.83            1            0            0            0       0.0001            0            0
    0.847831     0.387556    -0.361917  0.000316518      334.106      3135.94            1        5e-05 -1.08793e-11  3.39399e-13       0.0001 -4.35244e-11  1.35671e-12
```
The first line is the initial condition. It seems as if the quaternion state replaces the first four elements of the Cartesian state (even weirder the quaternion, if it is a quaternion, has a different value compared to the propagated one). 

Thank you,

Michele

**Comment by DominicDirkx** [comment]
Hi Michele,

That's indeed odd... I've been tinkering with the rotational dynamics code, mostly to include it into state estimation. I've pushed my current version of the code here:

https://github.com/DominicDirkx/tudat/tree/RotationalDynamicsEstimation

There are also one or two new tests in which rotation and translation are concurrently propagated. It may be that this update will fix the above problem. If you try it, make sure to do so on a separate branch, so that if the merge with your code goes horribly wrong somehow, you can easily return to your original code,

Dominic

**Comment by DominicDirkx** [comment]
Just realized I didn't answer your full question. 

The getEquationsOfMotionNumericalSolution is set in the  SingleArcDynamicsSimulator::integrateEquationsOfMotion function. This calls the convertNumericalStateSolutionsToOutputSolutions function of the dynamicsStateDerivative_ (type boost::shared_ptr< DynamicsStateDerivativeModel< TimeType, StateScalarType > >). The other input is the getEquationsOfMotionNumericalSolutionRaw, which is the 'unprocessed' states after propagation.

For each entry in the state map, teh convertToOutputSolution function of the DynamicsStateDerivativeModel is called, which in turn calls the convertToOutputSolution function of each state derivative model through:

```
                stateDerivativeModelsIterator_->second.at( i )->convertToOutputSolution(
                            internalSolution.segment(
                                currentStateIndices.at( i ).first, currentStateIndices.at( i ).second ), time,
                            outputState.block( currentStateIndices.at( i ).first, 0,
                                               currentStateIndices.at( i ).second, 1 ) );

```
where the output is stored in the corresponding blocks of the internalSolution vector.

Hope this helps!

**Comment by mfacchinelli** [comment]
Hi Dominic, 

I finally figured out the problem! (Well, at least on of the many :p) I think that the issue was in line 297 of `dynamicsStateDerivativeModel.h` (i.e., [here](https://github.com/Tudat/tudat/blob/e289a7b846eae4f91375e5d327192baa72d2848c/Tudat/Astrodynamics/Propagators/dynamicsStateDerivativeModel.h#L297)), where the size of the output vector was being assigned. The size was based on the size of the internal solution, but when using the exponential map, the size is one less of the quaternions, and thus it was overwriting some elements and going out of range. I now changed `internalSolution.rows( )` to `totalConventionalStateSize_`, which appears to be giving the right results. 

I guess that this was not a problem when using USM for translational propagation, since the internal solution is longer than the Cartesian solution. Anyway, I ran the unit tests and they are passed successfully. 

Oddly, when using quaternions and MRPs, it still outputs the strange result for the Cartesian (translational) state. Even weirder, is that with quaternions, it replaces the first 4 elements of the Cartesian state:
```
3.17537e+06  1.45172e+06 -1.35912e+06      1188.48       335.29      3134.83            1            0            0            0       0.0001            0            0
0.847831     0.387556    -0.361917  0.000316518      334.106      3135.94            1        5e-05 -1.08793e-11  3.39399e-13       0.0001 -4.35244e-11  1.35671e-12
```
whereas with MRPs, just the first three (where the first line is the conventional, i.e., quaternion, state and the second on the propagated, i.e., MRPs, state):
```
3.17537e+06  1.45172e+06 -1.35912e+06      1188.48       335.29      3134.83            1            0            0            0       0.0001            0            0
3.17537e+06  1.45172e+06 -1.35912e+06      1188.48       335.29      3134.83            0            0            0            0       0.0001            0            0

-2.26288e-07  -1.0344e-07  9.65964e-08      1184.89      334.106      3135.94            1        5e-05 -1.08793e-11  3.39399e-13       0.0001 -4.35244e-11  1.35671e-12
-2.26288e-07  -1.0344e-07  9.65964e-08      1184.89      334.106      3135.94      2.5e-05 -5.43966e-12    1.697e-13            0       0.0001 -4.35244e-11  1.35671e-12
```
I will work on this this afternoon, but it is very strange that it only appears for these two propagators and not for the exponential map. I guess it was to do with the size of the state somehow. 

Kind regards, 

Michele

**Comment by mfacchinelli** [comment]
One small update, it appears that the error is in the conversion, since the output of the integrator seems correct (i.e., the position is still in the order of 10<sup>6</sup>).

**Comment by mfacchinelli** [comment]
Ok, more updates. It would appear as if the Cartesian state is processed according to the normalizing function from the rotational dynamics. Hence, in case of quaternions, the first four elements are taken as a quaternion vector and normalized.

**Comment by DominicDirkx** [comment]
Hi Michele,

Strange, that should already be what it does: the (default) normalization function for Cartesian state does nothing. 

Try to add some couts to see which function exactly is being used to normalize the Cartesian state.

Dominic

**Comment by mfacchinelli** [comment]
Hi Dominic, 

Another quick update. It turns out that only the exponential map propagator failed. The problem is that when converting to quaternions, the first element of quaternions (the real part) is always positive. This leads to the problem that the quaternion history shows discontinuities (roughly speaking, it's like if the quaternion state were reflected around the x-axis). This is a problem in this unit test, since the rotational ephemeris is created via interpolation of the history, which leads to some weird results around these discontinuities. 

So, right now I created a new function to process the whole conventional state history, which only has a value for the exponential map (both translational and rotational). But maybe this is not the best way of solving this (especially since as I coded the program now, it only works if there is a discontinuity, otherwise it gives even weirder results). A better way of doing this is to probably introduce a new element in the exponential map, similar to the flag for the modified Rodrigues parameters. In this way, one can keep track of the shadow components, and simply multiply the quaternions by -1 where the flag is active. If you have any ideas, let me know.

Kind regards, 

Michele

**Comment by mfacchinelli** [comment]
Hi Dominic, 

Yet another update on the rotational dynamics. Last week I finished the implementation, and as I said in the previous post, I implemented the extra flag in the exponential map state. I also added a new function inside the `integrateEquationsOfMotion( )` of the `SingleArcDynamicsSimulator` class, called `processConventionalStateHistory( )` (which is a function of the `DynamicsStateDerivativeModel` class), which processes the conventional state based on the propagated state. Of course, this is only used if one uses the exponential map state for propagation of the rotational dynamics. 

Kind regards, 

Michele
